search-provider/eks-metadata-provider.c

```diff
@@ -0,0 +1,693 @@
+{
+  EksMetadataProvider *self = EKS_METADATA_PROVIDER (object);
+
+  switch (prop_id)
+    {
+    case PROP_APPLICATION_ID:
+      g_clear_pointer (&self->application_id, g_free);
+      self->application_id = g_value_dup_string (value);
+      break;
+
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+    }
+}
+
+static void
+eks_metadata_provider_finalize (GObject *object)
+{
+  EksMetadataProvider *self = EKS_METADATA_PROVIDER (object);
+
+  g_clear_pointer (&self->application_id, g_free);
+  g_clear_object (&self->skeleton);
+  g_clear_pointer (&self->translation_infos, g_hash_table_unref);
+
+  G_OBJECT_CLASS (eks_metadata_provider_parent_class)->finalize (object);
+}
+
+static void
+eks_metadata_provider_class_init (EksMetadataProviderClass *klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->get_property = eks_metadata_provider_get_property;
+  object_class->set_property = eks_metadata_provider_set_property;
+  object_class->finalize = eks_metadata_provider_finalize;
+
+  eks_metadata_provider_props[PROP_APPLICATION_ID] =
+    g_param_spec_string ("application-id", "Application Id", "Application Id",
+      "", G_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY | G_PARAM_STATIC_STRINGS);
+
+  g_object_class_install_properties (object_class,
+                                     NPROPS,
+                                     eks_metadata_provider_props);
+}
+
+typedef struct _MetadataQueryState {
+  EksMetadataProvider   *provider;
+  GDBusMethodInvocation *invocation;
+} MetadataQueryState;
+
+static MetadataQueryState *
+metadata_query_state_new (EksMetadataProvider *provider,
+                          GDBusMethodInvocation      *invocation)
+{
+  MetadataQueryState *state = g_new0 (MetadataQueryState, 1);
+  state->provider = provider;
+  state->invocation = g_object_ref (invocation);
+
+  return state;
+}
+
+static void
+metadata_query_state_free (MetadataQueryState *state)
+{
+  g_clear_object (&state->invocation);
+
+  g_free (state);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (MetadataQueryState,
+                               metadata_query_state_free);
+
+static void
+add_key_value_pair_to_variant (GVariantBuilder *builder,
+                               const char      *key,
+                               GVariant        *value)
+{
+  g_variant_builder_open (builder, G_VARIANT_TYPE ("{sv}"));
+  g_variant_builder_add (builder, "s", key);
+  g_variant_builder_add (builder, "v", value);
+  g_variant_builder_close (builder);
+}
+
+static gboolean
+gvalue_to_variant_internal (GValue              *value,
+                            const GVariantType  *expected_type,
+                            GVariant           **out_variant,
+                            GError             **error)
+{
+  g_return_val_if_fail (out_variant != NULL, FALSE);
+
+  /* Special case: if expected_type is G_VARIANT_TYPE_VARDICT then we need
+   * to serialize from JsonObject. */
```

Comment:
`serialize` -> `deserialize`, no?

Reply:
Yeah, I suppose when I wrote this I meant "serialize to GVariant from JsonObject" except that the relevant method is ends with `_deserialize`